mypy/checker.py

```diff
@@ -1480,7 +1480,7 @@ def count_nested_types(self, typ: Instance, check_type: str) -> int:
     def visit_yield_stmt(self, s: YieldStmt) -> Type:
         return_type = self.return_types[-1]
         if isinstance(return_type, Instance):
-            if return_type.type.fullname() != 'typing.Iterator':
+            if not is_subtype(return_type, self.named_type('typing.Iterable')):
```

Comment:
This has a few issues. Returning a user-defined type is questionable, as the concrete type is going to be `generator` always, and it may not have all the methods defined in the user-defined type. Also, line 1486 assumes that the type has exactly a single type parameter and this gets mapped to the type parameter of `Iterable`. 

I think that the only valid function return types should be these, and we can just special case them here:

* `typing.Iterator[...]`
* `typing.Generator[...]`
* `Any`